scio-core/src/main/scala/com/spotify/scio/coders/instances/AvroCoders.scala

```diff
@@ -127,25 +131,48 @@ trait AvroCoders {
   def avroGenericRecordCoder: Coder[GenericRecord] =
     Coder.beam(new SlowGenericRecordCoder)
 
+  // Try to get the schema with SpecificData.getSchema
+  // This relies on private SCHEMA$ field that may not be defined on custom SpecificRecord instance
+  // Otherwise create a default instance and call getSchema
+  private def schemaForClass[T <: SpecificRecord](clazz: Class[T]): Try[Schema] =
+    Try(SpecificData.get().getSchema(clazz))
+      .orElse(Try(clazz.getDeclaredConstructor().newInstance().getSchema))
+
   implicit def avroSpecificRecordCoder[T <: SpecificRecord: ClassTag]: Coder[T] = {
     val clazz = ScioUtil.classOf[T]
+    val schema = schemaForClass(clazz).getOrElse {
+      val msg =
+        "Failed to create a coder for SpecificRecord because it is impossible to retrieve an " +
+          s"Avro schema by instantiating $clazz. Use only a concrete type implementing " +
+          s"SpecificRecord or use GenericRecord type in your transformations if a concrete " +
+          s"type is not known in compile time."
+      throw new RuntimeException(msg)
+    }
+
+    val factory = new AvroDatumFactory(clazz) {
```

Comment:
I'm not sure we should provide this as default.

If user rely on AvroHugger, they should probably defined their own `AvroDatumFactory` as this one. I think now, most avro IOs will also require a custom `DatumReaderFactory/DatumWriterFactory` because of invalid schema loading.

Reply:
It was never possible to use the avro hugger generated classes in the scio `SpecificRecordIO` because beam automatically calls `ReflectData.get().getSchema(recordClass)` that throws an `AvroRuntimeException` [here](https://github.com/apache/beam/blame/6152a70d64b08809940ec07c8df2d4f0168d49ec/sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/io/AvroIO.java#L387)

Reply:
I'd be more in favor of this change for the avro coders: https://github.com/spotify/scio/pull/4928/commits/e1dd5ea4a51d4f10243756860d014b01af2301b0